File: Dec2020/11Dec/PolicemenAndThieves.cpp
/*
Problem Name: Policemen and thieves
Problem Link: https://www.hackerearth.com/practice/algorithms/searching/linear-search/practice-problems/algorithm/joker-and-thieves-53e59f4a/description/
*/

#include<bits/stdc++.h>
using namespace std;

#define test(t)     int t; cin>>t; while(t--)

int main(){
    ios::sync_with_stdio(0);    
    cin.tie(0);             
    test(t){
        int n, k, caught = 0;
        cin>>n>>k;
        char A[n][n];
        for(int i=0; i<n; ++i){
            for(int j=0; j<n; ++j){
                cin>>A[i][j];
            }
        }
        for(int i=0; i<n; ++i){
            vector<int> police;
            vector<int> thief;
            for(int j=0; j<n; ++j){
                if(A[i][j] == 'P')
                    police.push_back(j);
                else
                    thief.push_back(j);
            }
            int p = 0, t = 0;
            while(p < police.size() && t < thief.size()){
                if(abs(police[p] - thief[t]) <= k){
                    p++;
                    t++;
                    caught++;
                }
                else if(police[p] < thief[t])
                    p++;
                else
                    t++;
            }
            police.clear();
            thief.clear();
        }
        cout<<caught<<"\n";
    }
    return 0;
}

File: Dec2020/11Dec/PolicemenAndThieves.java
/*
Problem Name: Policemen and thieves
Problem Link: https://www.hackerearth.com/practice/algorithms/searching/linear-search/practice-problems/algorithm/joker-and-thieves-53e59f4a/
*/
import java.io.*;
import java.util.*;
public class TestClass {
    static int solution(char[][] A, int K){
        // Write your code here
        List<Integer> police = new ArrayList<Integer>();
        List<Integer> thief = new ArrayList<Integer>();
        int n = A.length, ans = 0;
        for(int i=0;i<n;i++){
            police.clear();
            thief.clear();
            for(int j=0;j<n;j++){
                if(A[i][j] == 'P')
                    police.add(j);
                else
                    thief.add(j);
            }
            int p = 0, t = 0;
            for(;(p < police.size()) && (t < thief.size());){
                if(Math.abs(police.get(p) - thief.get(t)) <= K){
                    ans++;
                    p++;
                    t++;
                }
                else if(police.get(p) < thief.get(t))
                    p++;
                else
                    t++;
            }
        }
        return ans;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter wr = new PrintWriter(System.out);
        int T = Integer.parseInt(br.readLine().trim());
        for(int t_i=0; t_i<T; t_i++)
        {
            String[] line = br.readLine().split(" ");
            int N = Integer.parseInt(line[0]);
            int K = Integer.parseInt(line[1]);
            char[][] A = new char[N][N];
            for(int i_A=0; i_A<N; i_A++)
            {
            	String[] arr_A = br.readLine().split(" ");
            	for(int j_A=0; j_A<arr_A.length; j_A++)
            	{
            		A[i_A][j_A] = arr_A[j_A].charAt(0);
            	}
            }

            int out_ = solution(A, K);
            System.out.println(out_);
            System.out.println("");
         }

         wr.close();
         br.close();
    }
}
